web-common/src/features/canvas/state-managers/state-managers.ts

```diff
@@ -39,18 +58,115 @@ export function removeCanvasStore(
   instanceId: string,
 ): void {
   const id = makeCanvasId(canvasName, instanceId);
-  canvasStores.delete(id);
+  canvasRegistry.delete(id);
 }
 
-export function createCanvasStore(
+export function setCanvasStore(
   canvasName: string,
   instanceId: string,
+  response: CanvasResponse,
 ): CanvasStore {
-  const canvasEntity = new CanvasEntity(canvasName, instanceId);
+  const id = makeCanvasId(canvasName, instanceId);
+
+  if (canvasRegistry.has(id)) {
+    console.warn(
+      `Canvas store for ID ${id} already exists. Returning existing store.`,
+    );
+    return canvasRegistry.get(id)!;
+  }
 
-  return {
+  const canvasEntity = new CanvasEntity(canvasName, instanceId, response);
+  const store: CanvasStore = {
     runtime: runtime,
     canvasEntity,
     queryClient,
   };
+
+  canvasRegistry.set(id, store);
+  return store;
+}
+
+const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const MAX_SPEC_RETRIES = 10;
+const RETRY_DELAY_MS = 1000;
+
+export async function handleCanvasStoreInitialization(
+  canvasName: string,
+  instanceId?: string,
+) {
+  let store: CanvasStore;
+
+  instanceId = instanceId ?? get(runtime).instanceId;
+
+  try {
+    store = getCanvasStore(canvasName, instanceId);
+    return {
+      store,
+      canvasName,
+    };
+  } catch {
+    //
+  }
+
+  try {
+    const queryOptions = getQueryServiceResolveCanvasQueryOptions(
+      instanceId,
+      canvasName,
+      {},
+    );
+
+    const fetchOptions = {
+      ...queryOptions,
+      retry: 10,
+    };
+
+    let data: V1ResolveCanvasResponse | null = null;
+    let validSpec: V1CanvasSpec | undefined = undefined;
+
+    for (let attempt = 1; attempt <= MAX_SPEC_RETRIES; attempt++) {
```

Comment:
Is this to get around currently deploying? The issue with this is we cannot show a loading spinner when we are retrying multiple times, there will just be a blank page.